File: ios/Sources/SegmentWriter.swift
import Foundation

/// PCM chunks into closed, capture-stamped WAV segments (stage C2's shadow,
/// mirroring the Android `SegmentWriter`): one minute of audio per file,
/// counted in bytes fed — audio time, not wall time — named
/// `<source>-YYYYMMDDTHHMMSS.wav` from this device's UTC clock at segment
/// open. WAV first for the same reason as Android: the delivery protocol is
/// container-agnostic and a platform encoder's header behaviour is probed
/// before it is trusted.
///
/// The mic path calls `offer` and returns; writing happens on the caller's
/// audio-callback-free queue (the stream's drainer feeds us, never the audio
/// thread directly). A segment never spans a disconnect gap: the client calls
/// `closeSegment()` when the stream drops, because the name claims continuity
/// from its stamp and the tee is gated on the connection.
final class SegmentWriter {
    static let sampleRate = 48_000
    static let segmentBytes = 60 * sampleRate * 2

    private let source: String
    private var handle: FileHandle?
    private var path: URL?
    private var written = 0
    var onSegmentClosed: (() -> Void)?

    init(source: String) {
        self.source = source
        SegmentStore.sweepOpen()
    }

    private static let stamp: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd'T'HHmmss"
        f.timeZone = TimeZone(identifier: "UTC")
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    static func wavHeader(dataBytes: Int) -> Data {
        var d = Data()
        func le32(_ v: UInt32) { withUnsafeBytes(of: v.littleEndian) { d.append(contentsOf: $0) } }
        func le16(_ v: UInt16) { withUnsafeBytes(of: v.littleEndian) { d.append(contentsOf: $0) } }
        d.append(contentsOf: Array("RIFF".utf8))
        le32(UInt32(36 + dataBytes))
        d.append(contentsOf: Array("WAVE".utf8))
        d.append(contentsOf: Array("fmt ".utf8))
        le32(16)
        le16(1)  // PCM
        le16(1)  // mono
        le32(UInt32(sampleRate))
        le32(UInt32(sampleRate * 2))
        le16(2)
        le16(16)
        d.append(contentsOf: Array("data".utf8))
        le32(UInt32(dataBytes))
        return d
    }

    func offer(_ data: Data) {
        var from = 0
        let bytes = [UInt8](data)
        while from < bytes.count {
            let out = handle ?? openSegment()
            guard let out else { return }
            let room = Self.segmentBytes - written
            let take = min(room, bytes.count - from)
            out.write(Data(bytes[from..<(from + take)]))
            written += take
            from += take
            if written >= Self.segmentBytes { closeSegment() }
        }
    }

    private func openSegment() -> FileHandle? {
        let name = "\(source)-\(Self.stamp.string(from: Date())).wav"
        let target = SegmentStore.open().appendingPathComponent(name)
        FileManager.default.createFile(atPath: target.path, contents: Self.wavHeader(dataBytes: 0))
        guard let out = try? FileHandle(forWritingTo: target) else { return nil }
        out.seekToEndOfFile()
        handle = out
        path = target
        written = 0
        return out
    }

    /// Patch the header with the truth and rename into the closed set — the
    /// only step anything downstream observes. Idempotent.
    func closeSegment() {
        guard let out = handle, let target = path else { return }
        handle = nil
        path = nil
        out.seek(toFileOffset: 0)
        out.write(Self.wavHeader(dataBytes: written))
        try? out.close()
        if written == 0 {
            try? FileManager.default.removeItem(at: target)
        } else {
            let dest = SegmentStore.root().appendingPathComponent(target.lastPathComponent)
            if (try? FileManager.default.moveItem(at: target, to: dest)) != nil {
                onSegmentClosed?()
            }
        }
        written = 0
    }
}

File: android/app/src/main/kotlin/org/recall/mic/SegmentWriter.kt
package org.recall.mic

import java.io.File
import java.io.RandomAccessFile
import java.time.Instant

/**
 * Turns the mic loop's PCM chunks into closed, capture-stamped WAV segments in
 * [SegmentStore] — the phone's half of store-and-forward
 * (recall/docs/architecture.md, stage C1), running in SHADOW: the PCM stream to
 * the recorder host is untouched, this only adds durable local copies that the
 * uploader delivers with verified receipts.
 *
 * The mic loop calls [offer], which hands bytes to a bounded [PcmSpool] and
 * returns — the same never-block-the-microphone rule as the network sender,
 * for the same measured reason. A writer thread drains the spool to disk and
 * rotates files every [SegmentNames.SEGMENT_BYTES] of audio (one minute of
 * PCM, counted in bytes: audio time, not wall time). Nothing here may throw
 * into the capture path; a dead writer costs the shadow copies, never the
 * stream.
 *
 * Pure JVM by construction (files, bytes, an injected clock), so rotation and
 * naming are unit-tested without a device.
 */
class SegmentWriter(
    private val base: File,
    private val source: String,
    private val now: () -> Instant = Instant::now,
    spoolBytes: Int = DEFAULT_SPOOL_BYTES,
    // Injected, not android.util.Log: the class stays pure JVM so its
    // rotation arithmetic is tested without a device; the service passes a
    // real logger.
    private val onFailure: (String) -> Unit = {},
    // Fires (on the writer thread) each time a segment lands in the closed
    // set — the service kicks the uploader here, so delivery tracks rotation
    // instead of waiting for the stream cycle to end.
    private val onSegmentClosed: () -> Unit = {},
) {
    private val spool = PcmSpool(spoolBytes)

    @Volatile private var running = false
    private var thread: Thread? = null

    private var file: RandomAccessFile? = null
    private var path: File? = null
    private var written = 0

    /** Never blocks; drops-oldest into the counted spool under pressure. */
    fun offer(chunk: ByteArray, length: Int) {
        spool.offer(chunk, length)
    }

    fun droppedBytes(): Long = spool.dropped()

    fun start() {
        if (running) return
        running = true
        // Whatever a crash left half-written is real audio: close it out first.
        SegmentStore.sweepOpen(base)
        thread =
            Thread({ drainLoop() }, "segment-writer").apply {
                isDaemon = true
                start()
            }
    }

    /** Flush, finalize the open segment, and stop. Idempotent. */
    fun close() {
        if (!running) return
        running = false
        thread?.join(JOIN_TIMEOUT_MS)
        thread = null
    }

    private fun drainLoop() {
        try {
            while (running) {
                val pending = spool.drain()
                if (pending.isEmpty()) {
                    Thread.sleep(IDLE_MS)
                } else {
                    write(pending)
                }
            }
            // One last drain so Stop does not orphan the tail of the spool.
            write(spool.drain())
        } catch (e: Exception) {
            // The shadow must never take the stream down with it.
            onFailure("segment writer failed: ${e.message}")
        } finally {
            runCatching { finishSegment() }
        }
    }

    private fun write(bytes: ByteArray) {
        var from = 0
        while (from < bytes.size) {
            val out = file ?: openSegment()
            val room = SegmentNames.SEGMENT_BYTES - written
            val take = minOf(room, bytes.size - from)
            out.write(bytes, from, take)
            written += take
            from += take
            if (written >= SegmentNames.SEGMENT_BYTES) finishSegment()
        }
    }

    private fun openSegment(): RandomAccessFile {
        val name = SegmentNames.segmentName(source, now())
        val target = File(SegmentStore.open(base), name)
        val out = RandomAccessFile(target, "rw")
        // Sizes lie zero until close; decoders read to EOF, so a crash keeps
        // a decodable file rather than headerless bytes.
        out.write(SegmentNames.wavHeader(0))
        file = out
        path = target
        written = 0
        return out
    }

    private fun finishSegment() {
        val out = file ?: return
        val target = path ?: return
        file = null
        path = null
        // Patch the header with the truth, land the bytes, then RENAME into
        // the closed set — the only step anything downstream can observe.
        out.seek(0)
        out.write(SegmentNames.wavHeader(written))
        out.fd.sync()
        out.close()
        if (written == 0) {
            // A zero-length segment says nothing; do not ship an empty claim.
            target.delete()
        } else if (target.renameTo(File(SegmentStore.root(base), target.name))) {
            onSegmentClosed()
        }
        written = 0
    }

    companion object {
        /** ~4 s of PCM headroom between mic and disk. */
        const val DEFAULT_SPOOL_BYTES =
            4 * SegmentNames.SAMPLE_RATE * SegmentNames.BYTES_PER_SAMPLE
        private const val IDLE_MS = 50L
        private const val JOIN_TIMEOUT_MS = 3_000L
    }
}
